lib/src/assists/dart/extract_variable_to_parameter.dart

```diff
@@ -0,0 +1,158 @@
+          paramList.parameters.any((e) => e.name?.lexeme == variableName);
+      if (paramAlreadyExists) return;
+
+      final positionalParamChangeBuilder = reporter.createChangeBuilder(
+        message: 'Extract variable as parameter',
+        priority: 79,
+      );
+
+      final namedParamChangeBuilder = reporter.createChangeBuilder(
+        message: 'Extract variable as named parameter',
+        priority: 80,
+      );
+
+      final variableTypeName =
+          node.declaredElement!.type.getDisplayString(withNullability: true);
+      final namedParams = paramList.parameters.where((e) => e.isNamed);
+      final positionalParams =
+          paramList.parameters.where((e) => e.isPositional);
+
+      positionalParamChangeBuilder.addDartFileEdit((builder) {
+        int offset;
+        var addPrecedingComma = false;
+        var addSucceedingComma = false;
+
+        if (positionalParams.isEmpty) {
+          offset = paramList.leftParenthesis.end;
+          addSucceedingComma = namedParams.isNotEmpty;
+        } else {
+          addPrecedingComma = true;
+          offset = positionalParams.last.endToken.next!.offset;
+        }
+
+        builder.addInsertion(offset, (builder) {
+          if (addPrecedingComma) builder.write(',');
+          builder.write('$variableTypeName $variableName');
+          if (addSucceedingComma) builder.write(',');
+        });
+
+        _deleteOriginalVariable(node, builder);
+      });
+
+      namedParamChangeBuilder.addDartFileEdit((builder) {
+        int offset;
+        bool addBracketPrecedingComma;
+        bool addEnclosingBrackets;
+        bool addPrecedingComma;
+
+        if (namedParams.isEmpty && positionalParams.isEmpty) {
+          offset = paramList.leftParenthesis.end;
+          addBracketPrecedingComma = false;
+          addEnclosingBrackets = true;
+          addPrecedingComma = false;
+        } else if (namedParams.isEmpty && positionalParams.isNotEmpty) {
+          final param = positionalParams.last;
+
+          offset = param.end;
+          addBracketPrecedingComma =
+              param.endToken.next!.type != TokenType.COMMA;
+          addEnclosingBrackets = true;
+          addPrecedingComma = false;
+        } else {
+          final param = namedParams.last;
+
+          offset = param.end;
+          addBracketPrecedingComma = false;
+          addEnclosingBrackets = false;
+          addPrecedingComma = param.endToken.next!.type != TokenType.COMMA;
+        }
+
+        builder.addInsertion(offset, (builder) {
+          if (addBracketPrecedingComma) builder.write(',');
+          if (addEnclosingBrackets) builder.write('{');
+          if (addPrecedingComma) builder.write(',');
+          if (!variableTypeName.endsWith('?')) builder.write('required ');
+          builder.write('$variableTypeName $variableName');
+          if (addEnclosingBrackets) builder.write('}');
+        });
+
+        _deleteOriginalVariable(node, builder);
+      });
```

Comment:
Similarly
```dart
void addNamedParameter(
  FormalParameterList parameterList,
  DartType parameterType,
  String parameterName,
) {
  final namedParams = parameterList.parameters.where((e) => e.isNamed);
  final positionalParams =
      parameterList.parameters.where((e) => e.isPositional);
  final isRequired = !parameterType.isNullable;

  switch ((positionalParams.isEmpty, namedParams.isEmpty)) {
    case (true, true):
      final offset = parameterList.leftParenthesis.end;
      addInsertion(offset, (builder) {
        builder.write('{');
        if (isRequired) builder.write('required ');
        builder.writeType(parameterType);
        builder.write(' $parameterName}');
      });
    case (false, true):
      final offset = positionalParams.last.end;
      final nextToken = positionalParams.last.endToken.next;
      final hasTrailingComma =
          nextToken != null && nextToken.type == TokenType.COMMA;

      if (hasTrailingComma) {
        addInsertion(nextToken.end, (builder) {
          builder.write(' {');
          if (isRequired) builder.write('required ');
          builder.writeType(parameterType);
          builder.write(' $parameterName,}');
        });
      } else {
        addInsertion(offset, (builder) {
          builder.write(', {');
          if (isRequired) builder.write('required ');
          builder.writeType(parameterType);
          builder.write(' $parameterName}');
        });
      }
    case (true, false) || (false, false):
      final offset = namedParams.last.end;
      addInsertion(offset, (builder) {
        builder.write(', ');
        if (isRequired) builder.write('required ');
        builder.writeType(parameterType);
        builder.write(' $parameterName');
      });
  }
}
```

Reply:
I think we are restricting the type a bit. What if I want to add a required non-null field?

Reply:
Do you mean required nullable parameter?

For simplicity, I am using `final isRequired = !parameterType.isNullable;` to decide whether to add the `required` keyword or not. We could definitely include an extra boolean parameter to support required nullable parameter.